tensorizer/serialization.py

```diff
@@ -3502,26 +3614,87 @@ def __init__(
         )
         self._write(self._file_header.to_bytes())
 
-        self._metadata_start = self._file.tell()
-        self._metadata_cur = (
-            self._metadata_start + 8
-        )  # position of next metadata entry we'd write. Leave 8 bytes for metadata length field
+        self._layout_pos: int = self._file.tell()
+        self._layout_size: int = self._write(_FileLayout.placeholder())
+        self._last_layout: Optional[_FileLayout] = None
+
+        self._flush()  # Flush potentially-buffered writes from `_write` calls
+
+        self._metadata_start: int = self._layout_pos + self._layout_size
+
+        # Offset of next metadata entry we'd write, relative to
+        # the metadata segment (i.e., bytes after self._metadata_start).
+        # Leave 8 bytes for metadata length field
+        self._metadata_offset: int = 8
+
+        # Next header write offset, relative to the header segment
+        # (i.e., bytes after self._metadata_end).
+        self._header_offset: int = 0
+
+        # Next tensor write position relative to the start of the data segment
+        # (i.e., bytes after self._header_end).
+        self._tensor_offset: int = 0
+
         self._metadata_end: Optional[int] = None
         self._header_end: Optional[int] = None
         if max_tensors:
-            # Estimate 256 bytes per metadata entry and 1024 bytes per header entry
-            self._metadata_end = self._metadata_cur + max_tensors * 256
-            # this is less about header_end itself but ensuring that tensor_start is on a 4096-byte aligned boundary
-            self._header_end = (
-                (self._metadata_end + max_tensors * 1024) + 4095
-            ) & ~4095
-
-        self._header_cur = (
-            self._metadata_end
-        )  # is the start of where to write header data, or None
-        self._tensor_cur = (
-            self._header_end
-        )  # is the start of where to write tensor data. or None
+            approx_metadata_size: int = 256 * max_tensors
+            approx_header_size: int = 1024 * max_tensors
+
+            self._metadata_end = self._metadata_start + 8 + approx_metadata_size
+            # Extend the metadata segment to end on a block boundary
+            # This allows later manipulation with the fallocate syscall
+            # on coöperating operating systems to extend this segment
+            # in the middle of the file
+            self._metadata_end -= self._metadata_end % -4096
```

Comment:
Should block boundaries be a constant? i.e. do we know that the aligned boundaries are the same on ARM vs x86-64?